### PR TITLE
RHMAP-19011 - Cordova Forms App crashes on iOS 11 when using a PhotoCapture field

### PR DESCRIPTION
**Pull request by camilamacedo86**
Task: https://issues.jboss.org/browse/RHMAP-19011

Following the steps to Test It:

1. Create a new project based on Form
2. Add a new client application by import
3. Import this branch 
4. Create a form with a photo capture field
5. deploy the form
6. Associate this for with the client app created based on this change
7. build the app for iOS and install on device
8. use the photo capture field to take a photo with the camera
9. select Use Photo and it should works 
10. Test the build and the camera for Android platform.

### CODE REVIEW

**Comment by camilamacedo86** [comment]
@davidffrench could you check it?

**Comment by jcesarmobile** [comment]
`NSLocationWhenInUseUsageDescription` is not really needed for the camera component, but it's needed for map o location fields, so LGTM.

I would use more personalized messages and not the ones from the example on the plugin README.md, this messages will be shown on the  app when it ask for permission.

File: config.xml [hunk 1]
```diff
@@ -36,4 +36,16 @@
     <plugin name="cordova-plugin-console" spec="1.0.3" />
     <plugin name="fh-cordova-plugin-device" spec="https://github.com/feedhenry/fh-cordova-plugin-device.git#fh0.0.4" />
     <plugin name="cordova-plugin-whitelist" spec="1.2.2" />
+    <edit-config target="NSCameraUsageDescription" file="*-Info.plist" mode="merge">
+        <string>need camera access to take pictures</string>
+    </edit-config>
+    <edit-config target="NSPhotoLibraryUsageDescription" file="*-Info.plist" mode="merge">
+        <string>need to photo library access to get pictures from there</string>
```
Comment (damienomurchu) [inline config.xml]: language grammar - just change from `need to photo library access to get pictures from there`to `need to access photo library to get pictures from there`

**Comment by jcesarmobile** [comment]
BTW, plugin versions here are old, those versions used variables to set this texts and the variables will overwrite the `edit-config` values.

This probably works because `NSPhotoLibraryAddUsageDescription` is new for iOS 11 so the plugins are not overwriting it,  but for the other entries, they are probably replaced by an empty string (default value). That's enough to make the app not crash, but the app won't be accepted if submitted to Itunes Connect as they don't accept empty strings.

So if you don't want to update the plugins, I would recommend to keep the `NSPhotoLibraryAddUsageDescription` entry and replace the others with something like:

```
<plugin name="cordova-plugin-camera" spec="2.3.0">
        <variable name="CAMERA_USAGE_DESCRIPTION" value="your usage message" />
        <variable name="PHOTOLIBRARY_USAGE_DESCRIPTION" value="your usage message" />
</plugin>
```

You have to do it for all plugins using this variables, because not setting one of them in a plugin can reset the string to the empty one if the plugin is installed after the one that set it in first place and I don't think you can be sure that a plugin will be installed before the other.

File: config.xml [hunk 2]
```diff
@@ -36,4 +36,16 @@
...
+        <string>need camera access to take pictures</string>
+    </edit-config>
+    <edit-config target="NSPhotoLibraryUsageDescription" file="*-Info.plist" mode="merge">
+        <string>need to access photo library to get pictures from there</string>
+    </edit-config>
+    <edit-config target="NSLocationWhenInUseUsageDescription" file="*-Info.plist" mode="merge">
+        <string>need location access to find things nearby</string>
+    </edit-config>
+    <edit-config target="NSPhotoLibraryAddUsageDescription" file="*-Info.plist" mode="merge">
+        <string>need to photo library access to save pictures there</string>
```
Comment (matskiv) [inline config.xml]: Same as Damien commented for NSCameraUsageDescription 
language grammar - just change from "need to photo library access to save pictures there" to "need to access photo library to save pictures there"

**Comment by matskiv** [comment]
Tested on iOS 10 and iOS11 devices. 👍

**Comment by camilamacedo86** [comment]
@jcesarmobile are you telling that for we move forward this code should be as following? 

```
    <plugin name="cordova-plugin-camera" spec="2.3.0">
        <variable name="CAMERA_USAGE_DESCRIPTION" value="your usage message" />
        <variable name="PHOTOLIBRARY_USAGE_DESCRIPTION" value="your usage message" />
    </plugin>
    <edit-config target="NSPhotoLibraryAddUsageDescription" file="*-Info.plist" mode="merge">
        <string>need to access photo library to save pictures there</string>
    </edit-config>
```

Is this the change that should be made here? 

@damienomurchu @matskiv the changes requests in the grammar/text are done. Is it ok? Or it should be made in another way?

**Comment by jcesarmobile** [comment]
Yes, but you have to do it for other plugins too, barcodescanner set the camera message, media-capture sets the camera and the photo library messages, if you only set it for camera plugin they might end with the empty string.

But right now all of them have the empty string, so it's not a big problem unless users want to submit the app to the App Store.

**Comment by camilamacedo86** [comment]
@jcesarmobile could you check now? Is it ok?

**Comment by jcesarmobile** [comment]
"your usage message" is just some example text, you have to change it with something that is relevant for the forms app. That text will be displayed when the app ask for permission. You have to give a reason why the app needs that permission.

In the last commit you have updated some of the plugins to the latest version, the latest version of the core plugins (camera, media, media-capture) don't use variables anymore, so you don't have to provide them, but use the edit-config as you did in the first commit. 

barcodescanner plugin doesn't need PHOTOLIBRARY_USAGE_DESCRIPTION, just the camera one.
media plugin doesn't need any of them.

**Comment by jcesarmobile** [comment]
media-capture is not latest, so it needs the variables or an update.
barcodescanner plugin still uses variables, so you have to set the variable for it (only needs the camera one)

If you go on with the plugins update, then this has to be properly tested on iOS and Android (build the apps, and do end to end testing for both of them) as major plugin updates might break things.

Also, it's probably safest to not use `~` in the plugin versions, patch updates should break the app, but who knows, better always use the version that was tested.

**Comment by camilamacedo86** [comment]
@jcesarmobile done 👍 
- Removed the `~`, build it in both platforms and add it to the steps to test here.

**Comment by jcesarmobile** [comment]
Why is it not compatible? Did you get any error? Are the other plugins you updated compatible?

About the PR, I still think that you should not use the exact example texts, you should use something more personalized. Anyway, before the PR it was just using the empty strings, so the change is an improvement whatsoever. 

Just FYI, if you have the variables, then the edit-config entries will be overwritten, so you don't need to add them. It's not a big deal as both have the same texts and at some point somebody might update the plugins and might forget to add the edit-config, so it's better to have them just in case. Also, geolocation plugin will overwrite the text to the empty string, the version it's using didn't even allow to use variables. Just telling you this in case you see an empty message and don't understand why it's not set if you provided the edit-config entry for it.

Aside from this comments, I think the PR is good and could be merged, but I don't have access to a build farm to test anymore, so I can't confirm

**Comment by camilamacedo86** [comment]
@jcesarmobile Tks for all support so far. Regards the compatibility of the plugin `cordova-plugin-media-capture` in this project, If we use the manager to install the plugin which I understand that is what is recommended then it will try to install the last compatible version. In this way, you can check that in a new Cordova app the latest version of this plugin is added automatically. 

```
Camila:myApp cmacedo$ cordova create myApp
Camila:studies cmacedo$ cd myApp/
Camila:myApp cmacedo$ cordova plugin add cordova-plugin-media --save
Fetching plugin "cordova-plugin-media@5.0.1" via npm
Saved plugin info for "cordova-plugin-media" to config.xml
Camila:myApp cmacedo$ cordova plugin add cordova-plugin-media-capture --save
Fetching plugin "cordova-plugin-media-capture@3.0.1" via npm
Saved plugin info for "cordova-plugin-media-capture" to config.xml

```

Into this project, the manager brings the latest compatible version with this project which is the version `1.4.0`.

**Comment by jcesarmobile** [comment]
Ah, I see what's going on, the problem is latest media-capture requires file plugin 5 or newer, but the app is using 4.2.0. The  file-transfer plugin used requires that version of file plugin, so there is a conflict. 
But you can't update file and file-transfer at the moment because of a bug already fixed but not released yet (it's going to be released in a few days)
So I would advise to just add this to fix the iOS 11 problem for now
```
<edit-config file="*-Info.plist" mode="merge" target="NSPhotoLibraryAddUsageDescription">
    <string>need to access photo library to save pictures there</string>
</edit-config>
```
And do the plugin update in the future if needed.

**Comment by jcesarmobile** [comment]
I *never* told you to update the plugins, I just said that they were old and used variables to set those messages, so your `edit-config` entries would be ignored. I think if you update them, you should update *all* of them. And right now there is a bug that won't allow you to update all of them because of what I told you in my previous message.

In your first commit you added multiple `edit-config` entries, that's not what I'm telling you to do, I'm telling you to just add the entry for `NSPhotoLibraryAddUsageDescription`, not the others you added on the first commit. The other messages will be set by the plugin variables with an empty message. The empty message is not recommended, but that's how it's working now. If you want to set those messages you have to use the variables for all the plugins that need them.

**Comment by camilamacedo86** [comment]
@jcesarmobile  done. Could you check if it is ok now?

**Comment by jcesarmobile** [comment]
Looks good to me

**Comment by camilamacedo86** [comment]
@jcesarmobile really tks for all your explanations, support and reviews here. Tks.

File: config.xml [hunk 3]
```diff
@@ -36,4 +36,7 @@
     <plugin name="cordova-plugin-console" spec="1.0.3" />
     <plugin name="fh-cordova-plugin-device" spec="https://github.com/feedhenry/fh-cordova-plugin-device.git#fh0.0.4" />
     <plugin name="cordova-plugin-whitelist" spec="1.2.2" />
-</widget>
+    <edit-config file="*-Info.plist" mode="merge" target="NSPhotoLibraryAddUsageDescription">
+        <string>need to access photo library to save pictures there</string>
```
Comment (matskiv) [inline config.xml]: I agree with Julio, this string could be more personalized/precise, perhaps: 
> Photo capture functionality needs access to photo library to save pictures.

@camilamacedo86  @davidffrench  Wdyt ?

**Comment by camilamacedo86** [comment]
@matskiv I don't have the authorization to merge the PR here. So, it need be done for other person or I need to be authorized.

**Comment by matskiv** [comment]
@camilamacedo86  I don't have write access too. And we both are in "Collaborators - Write Access" team, so there clearly is some misconfiguration. 
@davidffrench Do you have admin rights? Can you check permissions setup, please ? :)

**Comment by davidffrench** [comment]
@matskiv @camilamacedo86 I do have permission to merge this but i am unsure why ye do not. There are some steps that need to be taken before this is merged.

@camilamacedo86 Please update the status of the JIRA ticket to "Hand Over to QA"- https://issues.jboss.org/browse/RHMAP-19011.
@matskiv you can start start QA and should be assign yourself in the tester field. if you have verified it, you should also be the one who should move it to verified.

@camilamacedo86 
1. Please copy the verification steps to the JIRA ticket also. 
2. This change also needs to be cherry picked to the FH-v3.20 branch to get included in the next 3.x release.
3. Make sure that PR is also linked in the JIRA.

**Comment by camilamacedo86** [comment]
@davidffrench all done. However, I am still without the option/permission to do merges here. @matskiv I added you as a tester, could you just check the JIRA and move forward the status since you already tested it.

**Comment by camilamacedo86** [comment]
@matskiv I think the "Ready for QA" once it was merged" it is for ngui just because of the instance. For the libs, we don't have a QA so I understand that Is  "Ready for QA and Verify when it is merged.

**Comment by davidffrench** [comment]
@matskiv That is interesting. We will be doing it slightly differently on this team. All tickets (where appropriate) need to be verified by another member of the team before the PR is merged. This is because the PR may not actually resolve the problem. It needs to be verified before it is merged.

**Comment by matskiv** [comment]
@davidffrench  I agree that it is best to verify change before merging. I just was not used to transitioning JIRA into QE states before everything is merged. 
I was following this[1] definition for "Dev Complete" state (between PR and QE states), which basically says: "Everything needs to be merged to appropriate branches before becoming Dev Complete". It would be nice for us to document our process so everyone can read it anytime :).

[1] https://mojo.redhat.com/docs/DOC-1046609#jive_content_id_Dev_Complete

**Comment by camilamacedo86** [comment]
@davidffrench all PRs that I did so far work as @matskiv describe either. The verify status was just added after it is merged.

**Comment by matskiv** [comment]
Verified on Android 👍 

P.S: As for the process discussion, un-merged PRs were my main concern with current process.

**Comment by davidffrench** [comment]
@matskiv thanks for verifying. @camilamacedo86 once this is merged, can you please cherry pick and create a PR against the FH-v3.20 branch also, link that PR in the JIRA as well. Thank you

**Comment by camilamacedo86** [comment]
@davidffrench Ok. However, I am still without the authorization to merge it. So, how can we go forward?

**Pull request merged**